### PR TITLE
OpenKh.WinShell - IMDUtilities and IMZUtilities

### PR DESCRIPTION
**Pull request by Topaz-Reality**
**OpenKh.Winshell.IMDUtilities**

**What is this?**
WinShell.IMDUtilities is an implementation of IMD files into WinShell in an attempt to improve the user experience when working with IMD files.

**What can it do?**

Thumbnail Preview:
In the olden times in order to know what IMD file was, we had to open all of them, one by one, and see if they were the ones we were looking for. However, with WinShell.IMDUtilities installed, IMD files now generate thumbnails much like PNG files, which negate the need to go through every single file, making the chore of dealing with them a breeze.

Demo PNG: 
image

Batch Conversion to PNG:
Since we do not need advanced methods to get a PNG file out of an IMD, I figured to add a "Convert to PNG" option to the context menu to negate the need to open the tool up and fiddle with it if all we want to do is to get a PNG out of it, and this can be used in batch by selecting all IMD files in question and clicking the option in question.

Demo PNGs:
image
image

**Why?**
Because I am sick of going through a million IMDs just to get what I want.

### CODE REVIEW

File: OpenKh.Kh2/Ard/Event.cs [hunk 1]
```diff
@@ -856,8 +856,8 @@ private static void WriteVoices(MappingWriteArgs args)
 
         private static object ReadSetCameraData(MappingReadArgs args)
         {
-            List<SetCameraData.CameraKeys> AssignValues(SetCameraDataHeader header, IList<SetCameraData.CameraKeys> values) =>
```
Comment (Xeeynamo) [inline OpenKh.Kh2/Ard/Event.cs:859]: Why did you change the variable names here?

File: OpenKh.WinShell.IMDThumbnail/OpenKh.WinShell.IMDUtilities.csproj [hunk 2]
```diff
@@ -0,0 +1,79 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Xeeynamo) [inline OpenKh.WinShell.IMDThumbnail/OpenKh.WinShell.IMDUtilities.csproj:1]: This is the problem on why the build is failing. You are using the old csproj style, while you should use the new sdk csproj style. There are some tools that can convert it, but my strong suggestion to have the cleanest result is to backup all your files for this project BUT the csproj and `AssemblyInfo.cs`. Then create a new .NET Core project, put the files back and add again the missing NuGet packages and project references. The only difference is that you have to change the value of `<TargetFramework>netcoreapp3.1</TargetFramework>` as `net472`.

File: OpenKh.WinShell.IMDThumbnail/IMDThumbnail.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+using SharpShell.Attributes;
+using SharpShell.SharpThumbnailHandler;
+
+using OpenKh.Kh2;
+using OpenKh.Imaging;
+
+namespace OpenKh.WinShell.IMDUtilities
+{
+    [Obsolete]
```
Comment (Xeeynamo) [inline OpenKh.WinShell.IMDThumbnail/IMDThumbnail.cs:14]: Why is this marked as Obsolete?

File: OpenKh.WinShell.IMDThumbnail/MarshalBitmap.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            Height = Input.Height;
+            int num = bitmapData.Stride * Input.Height;
+            Bits = new int[num / 4];
+            Marshal.Copy(bitmapData.Scan0, Bits, 0, num / 4);
+            Input.UnlockBits(bitmapData);
+            BitsHandle = GCHandle.Alloc(Bits, GCHandleType.Pinned);
+            Bitmap = new Bitmap(Width, Height, Width * 4, PixelFormat.Format32bppArgb, BitsHandle.AddrOfPinnedObject());
+        }
+
+        public void SetPixel(int x, int y, Color colour)
```
Comment (Xeeynamo) [inline OpenKh.WinShell.IMDThumbnail/MarshalBitmap.cs:53]: As GetPixel and SetPixel are unused, I suggest to remove them

File: OpenKh.sln [hunk 5]
```diff
@@ -146,6 +146,9 @@ EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenKh.Research.Kh2Anim", "OpenKh.Research.Kh2Anim\OpenKh.Research.Kh2Anim.csproj", "{D4AF702B-145A-4052-B347-15166C43AB3F}"
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenKh.Command.TexFooter", "OpenKh.Command.TexFooter\OpenKh.Command.TexFooter.csproj", "{408B6E0A-20C1-4C55-A6BF-5342F2A16668}"
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "WinShell", "WinShell", "{FBD89BEB-3D49-440C-9A3C-1B52B5053DAD}"
```
Comment (Xeeynamo) [inline OpenKh.sln:149]: You are referencing a certain `WinShell` here, but I do not see it in the pull request

File: OpenKh.Kh2/TextureFooter/TextureFooterData.cs [hunk 6]
```diff
@@ -71,24 +71,24 @@ public TextureFooterData(Stream stream)
                             {
                                 stream.Position = subStreamPos + firstPosition;
                                 var indexedFrameList = new Dictionary<int, TextureFrame>();
-                                var index = 0;
```
Comment (Xeeynamo) [inline OpenKh.Kh2/TextureFooter/TextureFooterData.cs:74]: More unrelated changes

**Review by Xeeynamo — CHANGES_REQUESTED** [review]

File: OpenKh.Kh2/Ard/Event.cs (hunk 1 above)
Comment (Topaz-Reality) [reply OpenKh.Kh2/Ard/Event.cs:859]: Build was failing on NET Standard because of conflicting variable names.
"This variable already exists on this scope"

File: OpenKh.WinShell.IMDThumbnail/OpenKh.WinShell.IMDUtilities.csproj (hunk 2 above)
Comment (Topaz-Reality) [reply OpenKh.WinShell.IMDThumbnail/OpenKh.WinShell.IMDUtilities.csproj:1]: I see, I will do that in that case.

File: OpenKh.WinShell.IMDThumbnail/IMDThumbnail.cs (hunk 3 above)
Comment (Topaz-Reality) [reply OpenKh.WinShell.IMDThumbnail/IMDThumbnail.cs:14]: You see, because FileExtension is an obsolete feature but at the same time ClassOfExtension (which is meant to replace FileExtension) does not work with Thumbnail Generation.
This should be addressed in the next version of SharpShell.

File: OpenKh.WinShell.IMDThumbnail/MarshalBitmap.cs (hunk 4 above)
Comment (Topaz-Reality) [reply OpenKh.WinShell.IMDThumbnail/MarshalBitmap.cs:53]: Good Call.

File: OpenKh.sln (hunk 5 above)
Comment (Topaz-Reality) [reply OpenKh.sln:149]: That's the folder, I do not know why VS decided to make it a reference to a project.

File: OpenKh.Kh2/TextureFooter/TextureFooterData.cs (hunk 6 above)
Comment (Topaz-Reality) [reply OpenKh.Kh2/TextureFooter/TextureFooterData.cs:74]: Same issue as the first reply I made.

**Comment by Topaz-Reality** [comment]
I have addressed some of the stuff you requested. Could ya check to see if there are any issues?

**Comment by Topaz-Reality** [comment]
**Added: OpenKh.WinShell.IMZUtilities**

**What is this?**
WinShell.IMZUtilities is an implementation of IMZ files into WinShell in an attempt to improve the user experience when working with IMZ files.

**What can it do?**

Quick Repack:
By right-clicking one or more IMD files, the option to pack all selected IMD files into an IMZ file will now appear, creating an IMZ file with the name "output" into the same directory as the selected IMD files.

Demo PNGs:
image
image

Quick Extraction:
By right-clicking an IMZ file, now the option to extract it will appear. Once clicked, a folder will be made with the name of that IMZ in question and the IMD files within the IMZ will be extracted to the said folder.

Demo PNGs:
image
image

**Why?**
I figured, might as well.

File: OpenKh.WinShell.IMDUtilities/OpenKh.WinShell.IMDUtilities.csproj [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net472</TargetFramework>
+    <RootNamespace>OpenKh.WinShell.IMDUtilities</RootNamespace>
+    <AssemblyName>OpenKh.WinShell.IMDUtilities</AssemblyName>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
```
Comment (Xeeynamo) [inline OpenKh.WinShell.IMDUtilities/OpenKh.WinShell.IMDUtilities.csproj:9]: Is the `Condition` valid only for Release and only for AnyCPU wanted? Is it even required to output the assembly to `OpenKh.WinShell.Output`?

File: docs/winshell/INSTALL.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# OpenKH Documentation - WinShell Extensions
+In order to allow many quality of life improvements to aid in modding Kingdom Hearts games, OpenKH features several WinShell extensions to serve many different purposes.
```
Comment (Xeeynamo) [inline docs/winshell/INSTALL.md:2]: Sorry to be a pain, bu tbetween the header and the paragraph there must be a new line. I suggest to use Visual Studio Code to know where the warnings are.

File: docs/winshell/INSTALL.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+- OpenKh.WinShell.IMDUtilities -> Provides thumbnail generation for IMD files and quick conversion of them to PNG.
+- OpenKh.WinShell.IMZUtilities -> Provides ease of extraction and repack for IMZ files.
+
+# Installation
+All OpenKh.WinShell libraries are powered by SharpShell. So their installation is not standard, but is easy to do none-the-less. 
+To install these extensions, download the SharpShell ServerManager from [this link](https://github.com/dwmkerr/sharpshell/releases).
+
+Afterwards, open the ServerManager (you need Administrator Privileges for this). You will be greeted with this screen:
+
+![](https://i.imgur.com/VpByIAO.png)
```
Comment (Xeeynamo) [inline docs/winshell/INSTALL.md:15]: Between the square brackets, always add a brief description of what the image is about.
I also suggest to host those images in the docs rather than in a different domain.

**Review by Xeeynamo — CHANGES_REQUESTED** [review]

**Comment by Xeeynamo** [comment]
You might also want to edit `azure-pipelines.yml` and `build.ps1` to build every project under the name pattern `OpenKh.WinShell.*`. That is an hack to exclude certain projects from Linux and macOS builds.

File: OpenKh.WinShell.IMDUtilities/OpenKh.WinShell.IMDUtilities.csproj (hunk 7 above)
Comment (Topaz-Reality) [reply OpenKh.WinShell.IMDUtilities/OpenKh.WinShell.IMDUtilities.csproj:9]: Well, the reason for the output being WinShell.Output is to prevent massive DLL copying and avoiding many undeletable and duplicate services in installation. But I will edit the condition to be valid at all costs. I somehow missed that, whoops!

File: docs/winshell/INSTALL.md (hunk 8 above)
Comment (Topaz-Reality) [reply docs/winshell/INSTALL.md:2]: I see. I made this in Stackedit.IO because of it's preview function, will fix.

File: docs/winshell/INSTALL.md (hunk 9 above)
Comment (Topaz-Reality) [reply docs/winshell/INSTALL.md:15]: First point is understood.
GitHub didn't let me do the second one, but I shall try again.

**Comment by Topaz-Reality** [comment]
Ping for changes.

**Review by Xeeynamo — APPROVED** [review]

**Comment by Xeeynamo** [comment]
All good 👍 very exciting for this and what it will come in the near future.
